src/kernel/network.cpp

```diff
@@ -55,211 +147,183 @@ CryptoKernel::Network::Network(CryptoKernel::Log* log,
-                    removals.insert(it->first);
-                }
-            }
-
-            for(const auto& peer : removals) {
-                const auto it = connected.find(peer);
-                peers->put(dbTx.get(), peer, it->second->info);
-                if(it != connected.end()) {
-                    connected.erase(it);
-                }
-            }
-
-            for(const auto& peer : peerInfos) {
-                peers->put(dbTx.get(), peer.first, peer.second);
-            }
+void CryptoKernel::Network::makeOutgoingConnectionsWrapper() {
+	while(running) {
+		makeOutgoingConnections();
+		std::this_thread::sleep_for(std::chrono::milliseconds(100));
+	}
+}
 
-            dbTx->commit();
+void CryptoKernel::Network::infoOutgoingConnectionsWrapper() {
+	while(running) {
+		bool wait = false;
 
-            std::lock_guard<std::mutex> cLock(connectedStatsMutex);
-            connectedStats.clear();
+		infoOutgoingConnections();
+		std::this_thread::sleep_for(std::chrono::milliseconds(100));
+	}
+}
 
-            for(const auto& peer : connected) {
-                peerStats stats = peer.second->peer->getPeerStats();
-                stats.version = peer.second->info["version"].asString();
-                stats.blockHeight = peer.second->info["height"].asUInt64();
-                connectedStats.insert(std::make_pair(peer.first, stats));
-            }
-        }
+void CryptoKernel::Network::makeOutgoingConnections() {
+	std::map<std::string, Json::Value> peersToTry;
+	{
+		CryptoKernel::Storage::Table::Iterator* it = new CryptoKernel::Storage::Table::Iterator(
+				peers.get(), networkdb.get());
+
+		for(it->SeekToFirst(); it->Valid(); it->Next()) {
+			if(connected.size() >= 8) { // honestly, this is enough
+				std::this_thread::sleep_for(std::chrono::seconds(20)); // so stop looking for a while
+			}
+
+			Json::Value peerInfo = it->value();
+
+			if(connected.contains(it->key())) {
+				continue;
+			}
+
+			std::time_t result = std::time(nullptr);
+
+			const auto banIt = banned.find(it->key());
+			if(banIt != banned.end()) {
+				if(banIt->second > static_cast<uint64_t>(result)) {
+					continue;
+				}
+			}
+
+			if(peerInfo["lastattempt"].asUInt64() + 5 * 60 > static_cast<unsigned long long int>
+					(result) && peerInfo["lastattempt"].asUInt64() != peerInfo["lastseen"].asUInt64()) {
+				continue;
+			}
+
+			sf::IpAddress addr(it->key());
+
+			if(addr == sf::IpAddress::getLocalAddress()
+					|| addr == myAddress
+					|| addr == sf::IpAddress::LocalHost
+					|| addr == sf::IpAddress::None) {
+				continue;
+			}
+
+			log->printf(LOG_LEVEL_INFO, "Network(): Attempting to connect to " + it->key());
+			peersToTry.insert(std::pair<std::string, Json::Value>(it->key(), peerInfo));
+		}
+		delete it;
+	}
+
+	// here, we only access local data (except where there are more locks)
+	for(std::map<std::string, Json::Value>::iterator entry = peersToTry.begin(); entry != peersToTry.end(); ++entry) {
```

Comment:
Long-running threads like this need to check for `!running` and `break` so that `ckd` quits in a timely manner.